aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,602 @@
+        oH != gradOutput.size(dimh) ||
+        oW != gradOutput.size(dimw)) {
+      AT_ERROR(
+          "Inconsistent gradOutput size. oT= ",
+          oT,
+          ", oH= ",
+          oH,
+          ", oW= ",
+          oW,
+          ". gradOutput: ",
+          gradOutput.size(dimt),
+          "x",
+          gradOutput.size(dimh),
+          "x",
+          gradOutput.size(dimw));
+    }
+    AT_CHECK(
+        gradOutput.ndimension() == input.ndimension() &&
+            gradOutput.size(dimn) == nslices,
+        "gradOutput and input Tensors should have same number of dimensions and also the same number of channels/slices");
+  }
+}
+
+Tensor& max_unpooling3d_forward_out_cpu(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  int64_t oT = output_size[0];
+  int64_t oW = output_size[1];
+  int64_t oH = output_size[2];
+  max_unpooling3d_shape_check(
+      self, Tensor(), indices, output_size, stride, padding);
+
+  if (self.ndimension() == 5) {
+    output.resize_({self.size(0),
+                    self.size(1),
+                    oT,
+                    oH,
+                    oW});
+  } else {
+    output.resize_(
+        {self.size(0), oT, oH, oW});
+  }
+  output.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES(
+      self.type(), "max_unpooling3d_forward_out_cpu_frame", ([&] {
+        max_unpooling3d_forward_out_cpu_frame<scalar_t>(
+            output,
+            self.contiguous(),
+            indices.contiguous(),
+            oT,
+            oW,
+            oH,
+            stride[0],
+            stride[1],
+            stride[2],
+            padding[0],
+            padding[1],
+            padding[2]);
+      }));
+  return output;
+}
+
+Tensor max_unpooling3d_forward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  auto output = at::empty({0}, self.options());
+  max_unpooling3d_forward_out_cpu(
+      output, self, indices, output_size, stride, padding);
+  return output;
+}
+
+template <typename scalar_t>
+static void max_unpooling2d_backward_out_cpu_frame(
+    scalar_t* gradInput_p,
+    scalar_t* gradOutput_p,
+    int64_t* ind_p,
+    int64_t nslices,
+    int64_t iwidth,
+    int64_t iheight,
+    int64_t owidth,
+    int64_t oheight) {
+  bool has_error = false;
+  int64_t error_index = 0;
+  int k;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++) {
+    scalar_t* gradInput_p_k = gradInput_p + k * iwidth * iheight;
+    scalar_t* gradOutput_p_k = gradOutput_p + k * owidth * oheight;
+    int64_t* ind_p_k = ind_p + k * iwidth * iheight;
+
+    int64_t i, j;
+    int64_t maxp;
+
+    for (i = 0; i < iheight; i++) {
+      for (j = 0; j < iwidth; j++) {
+        maxp = ind_p_k[i * iwidth + j]; /* retrieve position of max */
+        if (maxp < 0 || maxp >= owidth * oheight) {
```

Comment:
You lost #pragma omp critical here

Reply:
(OK, to be fair, there never was a critical here to begin with, but if you're going to hit the same memory location from all threads, you should mark it as critical.)